Fix enter playmode exception caused by search

rfadeev opened the pull request:
### Summary
Fix null reference exception thrown when entering play mode with non-zero search hits.

### Details
Explanation of the root cause can be found in https://github.com/sabresaurus/PlayerPrefsEditor/issues/5#issuecomment-777046984 and in the commit e60aaf0bfe1a5b0ba0acfbbe21be3dec6eb2fcd9 message. Unity does serialization of `PlayerPrefsEditor` since it inherits `EditorWindow`.

### How to test
Checkout target branch. Open editor, search "unity", make sure search results are not empty. Enter playmode and observe exception in the console.

Checkout source branch. Repeat steps and observe no exception. Check that search results are persistent when stopping play mode. Check editor behaves as before when opening and closing.

Tested with 2019.4.8f1 on Windows.

Would be great if @tracer8 could test source branch as well.

Closes #5

rfadeev committed:
Extract prefs cache deserialization into separate method

rfadeev committed:
Fix exception when entering playmode with nonzero search hits

Exception was caused by PlayerPrefsEditor.filteredPlayerPrefs serialization. It is a list which stores items of PlayerPrefPair type which is has Serializable attribute. Since PlayerPrefPair.Value type is object, Unity failed to serialize that field when entering playmode. Then later the value was set to null after deserialization once playmode has started. The issue happened only if filteredPlayerPrefs was not empty.

Calling DeserializePrefsIntoCache() sets value of deserializedPlayerPrefs and UpdateSearch() sets filteredPlayerPrefs using deserializedPlayerPrefs  value.

Initial idea for fix was to implement ISerializationCallbackReceiver and call those methods in OnAfterDeserialize. That did not work as Unity throws UnityException about accessing PlayerSettings.companyName during serialization.

rfadeev committed:
Remove Serializable attribute from PlayerPrefPair

No need to serialize type as prefs lists are set in OnEnable now.

sabresaurus reviewed (changes requested):
This PR does not work on Mac due to the refactor resulting in calling GUILayout outside of OnGUI

> ArgumentException: You can only call GUI functions from inside OnGUI.
UnityEngine.GUIUtility.CheckOnGUI () (at /Users/bokken/buildslave/unity/build/Modules/IMGUI/GUIUtility.cs:311)
UnityEngine.GUI.get_skin () (at /Users/bokken/buildslave/unity/build/Modules/IMGUI/GUI.cs:42)
UnityEngine.GUILayout.Label (System.String text, UnityEngine.GUILayoutOption[] options) (at /Users/bokken/buildslave/unity/build/Modules/IMGUI/GUILayout.cs:9)
Sabresaurus.PlayerPrefsEditor.PlayerPrefsEditor.DeserializePrefsIntoCache () (at Assets/PlayerPrefsEditor-1.2.0/Editor/PlayerPrefsEditor.cs:1095)
Sabresaurus.PlayerPrefsEditor.PlayerPrefsEditor.OnEnable () (at Assets/PlayerPrefsEditor-1.2.0/Editor/PlayerPrefsEditor.cs:122)

rfadeev commented:
Thanks for checking. I see the problem. For macOS before in `OnGUI` both `deserializedPlayerPrefs` was assigned and `GUILayout.Label` was added based on last write time of plist file. I need to think how to move that back to `OnGUI` since drawing depends on plist file path.

rfadeev committed:
Set deserialized player prefs directly

Since for macOS cannot call DeserializePrefsIntoCache from OnEnable due to editor GUI function is called for macOS to draw plist last write time.

sabresaurus commented:
Hey @rfadeev looks good, thanks!

rfadeev commented:
@sabresaurus You are welcome 😊
